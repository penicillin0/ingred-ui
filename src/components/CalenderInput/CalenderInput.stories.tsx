import * as React from "react";
import CalenderInput from "./CalenderInput";
import { Story } from "@storybook/react/types-6-0";

export default {
  title: "Components/Inputs/CalenderInput",
  component: CalenderInput,
  args: {
    placeholder: "placeholder",
  },
};

export const Example: Story = () => <CalenderInput />;
