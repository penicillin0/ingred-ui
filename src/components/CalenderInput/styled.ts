import styled, { DefaultTheme } from "styled-components";

const setBorderColor = (
  theme: DefaultTheme,
  isFocused: boolean,
  isError: boolean,
) => {
  if (isError) return theme.palette.danger.main;
  else if (isFocused) return theme.palette.primary.main;
  else return theme.palette.divider;
};

export const Container = styled.div<{ isFocused: boolean; isError: boolean }>`
  display: "flex";
  width: 140px;
  padding: 8px 8px;
  font-size: 14px;
  color: ${({ theme, isError }) =>
    isError ? theme.palette.danger.main : theme.palette.black};
  background-color: ${({ theme }) => theme.palette.background.default};
  border: 1px solid ${({ theme }) => theme.palette.divider};
  border-radius: ${({ theme }) => theme.radius}px;
  border-color: ${({ theme, isFocused, isError }) =>
    setBorderColor(theme, isFocused, isError)};
`;

export const Input = styled.input<{ letterCount: number; isError: boolean }>`
  text-align: right;
  border: none;
  width: ${({ letterCount }) => 8 * letterCount}px;
  padding: 2px 0px;
  color: ${({ theme, isError }) =>
    isError ? theme.palette.danger.main : theme.palette.black};
  &:focus {
    background-color: ${({ theme }) => theme.palette.primary.main};
    color: ${({ theme }) => theme.palette.white};
    outline: none;
  }
`;
