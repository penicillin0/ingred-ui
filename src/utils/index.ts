export { deepmerge } from "./deepmerge";
export { createChainedFunction } from "./createChainedFunction";
export { getShadow } from "./getShadow";
export { hexToRgba } from "./hexToRgba";
export { getDuration } from "./reactTransitionGroup";
export { renderWithThemeProvider } from "./renderWithThemeProvider";
export { addScrollbarProperties } from "./scrollbar";
export { spacer, getMargin, getPadding } from "./spacer";
