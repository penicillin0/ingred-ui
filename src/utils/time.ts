import dayjs from "dayjs";
import moment from "moment";

// moment.jsで動いているreact-datesと互換性を持たせるためのメソッド
// DatePickerとDateRangePickerのリニューアルが完了したら除去する
export function dayjsToMoment(date: dayjs.Dayjs | null): moment.Moment | null {
  if (!date) return null;
  const dateString = date.format();
  return moment(dateString);
}

export function momentToDayjs(date: moment.Moment | null): dayjs.Dayjs | null {
  if (!date) return null;
  const dateString = date.format();
  return dayjs(dateString);
}
