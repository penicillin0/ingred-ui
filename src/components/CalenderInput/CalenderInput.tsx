import React, { useState, useRef } from "react";
import * as Styled from "./styled";
import dayjs from "dayjs";

const CalenderInput: React.FC = () => {
  const today = dayjs();
  const [year, setYear] = useState<number>(today.year());
  const [month, setMonth] = useState<number>(today.month() + 1);
  const [date, setDate] = useState<number>(today.date());

  const yearInput = useRef<HTMLInputElement>(null);
  const monthInput = useRef<HTMLInputElement>(null);
  const dateInput = useRef<HTMLInputElement>(null);

  const handleOnChange =
    (datePart: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
      if (datePart === "year") setYear(Number(event.target.value));
      if (datePart === "month") setMonth(Number(event.target.value));
      if (datePart === "date") setDate(Number(event.target.value));
    };

  const handleOnKeyPress =
    (datePart: string) => (event: React.KeyboardEvent<HTMLInputElement>) => {
      if (event.key === "Enter" || event.key === "ArrowRight") {
        if (datePart === "year") {
          if (!yearInput.current || !yearInput.current.selectionStart) return;
          if (yearInput.current.selectionStart === 4)
            monthInput.current?.focus();
        }
        if (datePart === "month") {
          if (!monthInput.current || !monthInput.current.selectionStart) return;
          if (monthInput.current.selectionStart === 2)
            dateInput.current?.focus();
        }
      }
      if (event.key === "ArrowLeft") {
        if (datePart === "month") {
          if (!monthInput.current || !monthInput.current.selectionStart) return;
          if (monthInput.current.selectionStart === 1)
            yearInput.current?.focus();
        }
        if (datePart === "date") {
          if (!dateInput.current || !dateInput.current.selectionStart) return;
          if (dateInput.current.selectionStart === 1)
            monthInput.current?.focus();
        }
      }
      if (event.key === "ArrowUp") {
        if (datePart === "year") setYear(year + 1);
        if (datePart === "month" && month < 12) setMonth(month + 1);
        if (datePart === "date" && date < 31) setDate(date + 1);
      }

      if (event.key === "ArrowDown") {
        if (datePart === "year") setYear(year - 1);
        if (datePart === "month" && month > 1) setMonth(month - 1);
        if (datePart === "date" && date > 1) setDate(date - 1);
      }
    };

  const handleInputNumberOnly = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    event.target.value = event.target.value.replace(/[^0-9]+/i, "");
  };

  return (
    <Styled.Container>
      <Styled.Input
        ref={yearInput}
        letterCount={4}
        type={"text"}
        aria-label="Year"
        value={year}
        onChange={handleOnChange("year")}
        onKeyDown={handleOnKeyPress("year")}
        onInput={handleInputNumberOnly}
      />
      /
      <Styled.Input
        ref={monthInput}
        letterCount={2}
        type={"text"}
        aria-label="Month"
        value={("00" + month).slice(-2)}
        onChange={handleOnChange("month")}
        onKeyDown={handleOnKeyPress("month")}
        onInput={handleInputNumberOnly}
      />
      /
      <Styled.Input
        ref={dateInput}
        letterCount={2}
        type={"text"}
        aria-label="Date"
        value={("00" + date).slice(-2)}
        onChange={handleOnChange("date")}
        onKeyDown={handleOnKeyPress("date")}
        onInput={handleInputNumberOnly}
      />
    </Styled.Container>
  );
};

export default CalenderInput;
