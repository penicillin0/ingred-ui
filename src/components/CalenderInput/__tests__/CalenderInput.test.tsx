import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import CalenderInput from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("CalenderInput component testing", () => {
  afterEach(cleanup);

  test("CalenderInput", () => {
    const { asFragment } = renderWithThemeProvider(<CalenderInput />);
    expect(asFragment()).toMatchSnapshot();
  });
});
