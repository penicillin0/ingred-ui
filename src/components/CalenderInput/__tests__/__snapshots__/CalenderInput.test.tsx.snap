// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalenderInput component testing CalenderInput 1`] = `
<DocumentFragment>
  <div
    class="sc-bczRLJ hOtffT"
  >
    <input
      aria-label="Year"
      class="sc-gsnTZi fKrwos"
      type="text"
      value="2022"
    />
    /
    <input
      aria-label="Month"
      class="sc-gsnTZi iVPPZG"
      type="text"
      value="12"
    />
    /
    <input
      aria-label="Date"
      class="sc-gsnTZi iVPPZG"
      type="text"
      value="12"
    />
  </div>
</DocumentFragment>
`;
